include/neighbor.h

```diff
@@ -30,40 +31,99 @@ namespace diskann {
     }
   };
 
-  static inline unsigned InsertIntoPool(Neighbor *addr, unsigned K,
-                                        Neighbor nn) {
-    // find the location to insert
-    unsigned left = 0, right = K - 1;
-    if (nn < addr[left]) {
-      memmove((char *) &addr[left + 1], &addr[left], K * sizeof(Neighbor));
-      addr[left] = nn;
-      return left;
+  // Invariant: after every `insert` and `pop`, `cur_` points to
+  //            the first Neighbor which is unchecked.
+  class NeighborSet {
+   public:
+
+    NeighborSet() : size_(0), capacity_(0), cur_(0) {
+    }
+
+    explicit NeighborSet(size_t capacity)
+        : size_(0), capacity_(capacity), cur_(0), data_(capacity_ + 1) {
+    }
+    
+    // Inserts the item ordered into the set up to the sets capacity.
+    // The item will be dropped if it is the same id as an exiting 
+    // set item or it has a greated distance than the final
+    // item in the set. The set cursor that is used to pop() the 
+    // next item will be set to the lowest index of an uncheck item 
+    void insert(const Neighbor &nbr) {
```

Comment:
Let's leave this as is now and revisit the use of the vector type. I don't think there is much benefit except for memory management when using the vector internally. This could be implemented using a simple array of Neighbor's